website/source/guides/writing-custom-terraform-providers.html.md

```diff
@@ -0,0 +1,553 @@
+
+_Partial mode_ is a mode that can be enabled by a callback that tells Terraform
+that it is possible for partial state to occur. When this mode is enabled, the
+provider must explicitly tell Terraform what is safe to persist and what is not.
+
+Here is an example of a partial mode with an update function:
+
+```go
+func resourceServerUpdate(d *schema.ResourceData, m interface{}) error {
+	// Enable partial state mode
+	d.Partial(true)
+
+	if d.HasChange("address") {
+		// Try updating the address
+		if err := updateAddress(d, meta); err != nil {
+			return err
+		}
+
+		d.SetPartial("address")
+	}
+
+	// If we were to return here, before disabling partial mode below,
+	// then only the "address" field would be saved.
+
+	// We succeeded, disable partial mode. This causes Terraform to save
+	// save all fields again.
+	d.Partial(false)
+
+	return nil
+}
+```
+
+Note - this code will not compile since there is no `updateAddress` function.
+You can implement a dummy version of this function to play around with partial
+state. For this example, partial state does not mean much in this documentation
+example. If `updateAddress` were to fail, then the address field would not be
+updated.
+
+## Implementing Destroy
+
+The `Destroy` callback is exactly what it sounds like - it is called to destroy
+the resource. This operation should never update any state on the resource. It
+is not necessary to call `d.SetId("")`, since any non-error return value assumes
+the resource was deleted successfully.
+
+```go
+func resourceServerDelete(d *schema.ResourceData, m interface{}) error {
+	d.SetId("")
+	return nil
+}
+```
+
+The destroy function should always handle the case where the resource might
+already be destroyed (manually, for example). If the resource is already
+destroyed, this should not return an error. This allows Terraform users to
+manually delete resources without breaking Terraform.
+
+```shell
+$ go build -o terraform-plugin-example
+```
+
+Run `terraform destroy` to destroy the resource.
+
+```text
+$ terraform destroy
+Do you really want to destroy?
+  Terraform will delete all your managed infrastructure.
+  There is no undo. Only 'yes' will be accepted to confirm.
+
+  Enter a value: yes
+
+example_server.my-server: Refreshing state... (ID: 5.6.7.8)
+example_server.my-server: Destroying... (ID: 5.6.7.8)
+example_server.my-server: Destruction complete
+
+Destroy complete! Resources: 1 destroyed.
+```
+
+## Implementing Read
+
+The `Read` callback is used to sync the local state with the actual state
+(upstream). This is called at various points by Terraform and should be a
+read-only operation. This callback should never modify the real resource.
+
+If the ID is updated to blank, this tells Terraform the resource no longer
+exists (maybe it was destroyed out of band). Just like the destroy callback, the
+`Read` function should gracefully handle this case.
+
+```go
+// Tells Terraform the resource no longer exists
+d.SetId("")
```

Comment:
Maybe explain that we need to READ the resource from the API, and update the values, so we detect drift and Terraform will alert. Maybe not need to show code, maybe just link out to an example. One of READs most important parts is that drift detection. 

ALSO, we have an EXISTS method that I think we consider best practice, who's only job is to determine if the resource still exists and if not to remove it with `SetId("")`. I say "consider" best practice because I think we do, but it's not actually widely implemented